src/vscode-bicep/src/feedback/surveys.ts

```diff
@@ -0,0 +1,338 @@
+    }
+  );
+}
+
+export interface ISurveyInfo {
+  akaLinkToSurvey: string; // the part after https://aka.ms/
+  surveyStateKey: string; // Key used in global storage to persist the survey state
+  postponeAfterTakenInDays: number;
+  postponeForLaterInDays: number;
+  surveyPrompt: string;
+}
+
+export interface ISurveyState {
+  lastTaken?: Date;
+  postponedUntil?: Date;
+}
+
+export interface IPersistedSurveyState {
+  lastTakenMs?: number;
+  postponedUntilMs?: number;
+}
+
+export class Survey {
+  public constructor(
+    private readonly globalState: GlobalState,
+    private readonly surveyInfo: ISurveyInfo,
+    private inject: {
+      showInformationMessage: typeof window.showInformationMessage;
+      getIsSurveyAvailable: typeof Survey.getIsSurveyAvailable;
+      launchSurvey: typeof Survey.launchSurvey;
+    } = {
+        showInformationMessage: window.showInformationMessage,
+        getIsSurveyAvailable: Survey.getIsSurveyAvailable,
+        launchSurvey: Survey.launchSurvey,
+      }
+  ) {
+    // noop
+  }
+
+  /**
+   * Shows the survey if it's available and timely, and the user doesn't opt out.
+   */
+  public async checkShowSurvey(
+    context: IActionContext,
+    now: Date
+  ): Promise<void> {
+    context.errorHandling.suppressDisplay = true;
+    context.telemetry.properties.isActivationEvent = "true";
+    context.telemetry.properties.akaLink = this.surveyInfo.akaLinkToSurvey;
+
+    const surveyState = this.getPersistedSurveyState(context, now);
+
+    const shouldShowSurvey = await this.shouldAskToTakeSurvey(
+      context,
+      surveyState,
+      now
+    );
+
+    if (shouldShowSurvey) {
+      await this.askToTakeSurvey(context, surveyState, now);
+    }
+
+    await this.updatePersistedSurveyState(context, surveyState);
+  }
+
+  private static getFullSurveyLink(akaLink: string): string {
+    return `https://aka.ms/${akaLink}`;
+  }
+
+  private async shouldAskToTakeSurvey(
+    context: IActionContext,
+    state: ISurveyState,
+    now: Date
+  ): Promise<boolean> {
+    {
+      const neverShowSurveys = this.getShouldNeverShowSurveys();
+      context.telemetry.properties.neverShowSurvey = String(neverShowSurveys);
+      if (neverShowSurveys) {
+        context.telemetry.properties.shouldAsk = "never";
+        return false;
+      }
+
+      context.telemetry.properties.lastTaken = state.lastTaken?.toUTCString();
+      context.telemetry.properties.postonedUntil =
+        state.postponedUntil?.toUTCString();
+
+      if (
+        state.postponedUntil &&
+        state.postponedUntil.valueOf() > now.valueOf()
+      ) {
+        context.telemetry.properties.shouldAsk = "postponed";
+        return false;
+      }
+
+      if (state.lastTaken) {
+        const okayToAskAgainMs =
+          state.lastTaken.valueOf() +
+          daysToMs(this.surveyInfo.postponeAfterTakenInDays);
+        if (okayToAskAgainMs > now.valueOf()) {
+          context.telemetry.properties.shouldAsk = "alreadyTaken";
```

Comment:
Do we really want to track this? This might be a very noisy metric between selecting "Sure" and until next time.

Reply:
Mixed feelings, but:
1) it will only run once per vs code session, just like the start-up one, and much less noisy than file open.  I can cut it down later, though, if it makes sense.  I'd rather have it in for now so we can verify things are happening correctly.
2) Right now, I don't think it ever gets sent if they don't select a button (although I'd call this a design flag in the package we use)